### ISSUE TITLE
stop_consuming (and disconnect) freeze

### ISSUE THREAD

**Issue by Hadatko** [eclipse-paho/paho.mqtt.cpp#410]
Hello, i couldn't find solution for my issue. 

Situation: In one thread i am using this function: messagePtr = client->consume_message();
I want unblock this function at the end of app (proper deinitialization) so i was thinking i can call disconnect/stop_consuming from other thread to unblock it and clean all  allocations. 

Is my expectation wrong and instead of that i need close consume_message somehow differently and then disconnect?

Hmm i tried workround it and stil disconnect is freezing.

<img width="456" alt="image" src="https://user-images.githubusercontent.com/13130700/194885566-18e17f76-1bec-42a2-91e8-6812196ce428.png">

**Comment by fpagliughi** [comment]
There should be a cleaner way to do this, supported by the library.

**Comment by fpagliughi** [comment]
Sorry I let this one slip by in the last release.

Clearly, calling `disconnect()` or `stop_consuming()` should push an empty message pointer into the queue to signal that a shutdown of one form or another is happening. That's what incoming "connection lost" and "disconnect" callbacks do internally.

I can add that to the v1.4.x line...

The upcoming v1.5 release will upgrade the "message" consumer into a more fully-functioning "event" consumer. With the event consumer, the existing calls like  `client->consume_message()` still work the same, but there will now also be calls like `client->consume_event()`. They will return a composite object that could be an incoming message or an event object to indicate "(re)connected", "connection lost", "disconnected", etc. These could have data, so the v5 disconnected event would contain the reason code and any properties from the DISCONNECT packet.

Basically, this all would allow you to just use the consumer and not need to install/use callbacks to get additional state notifications.

I guess, for this I should consider one or two more events: "stopped" if another local thread called `stop_consuming()`, and "disconnecting" if another local thread called `disconnect()`.

**Comment by RainbowSea5** [comment]
> Sorry I let this one slip by in the last release.
> 
> Clearly, calling `disconnect()` or `stop_consuming()` should push an empty message pointer into the queue to signal that a shutdown of one form or another is happening. That's what incoming "connection lost" and "disconnect" callbacks do internally.
> 
> I can add that to the v1.4.x line...
> 
> The upcoming v1.5 release will upgrade the "message" consumer into a more fully-functioning "event" consumer. With the event consumer, the existing calls like `client->consume_message()` still work the same, but there will now also be calls like `client->consume_event()`. They will return a composite object that could be an incoming message or an event object to indicate "(re)connected", "connection lost", "disconnected", etc. These could have data, so the v5 disconnected event would contain the reason code and any properties from the DISCONNECT packet.
> 
> Basically, this all would allow you to just use the consumer and not need to install/use callbacks to get additional state notifications.
> 
> I guess, for this I should consider one or two more events: "stopped" if another local thread called `stop_consuming()`, and "disconnecting" if another local thread called `disconnect()`.
This problem seems to have not been solved yet. I am using the code from branch 1.4. x, but even after the client disconnects voluntarily, it still blocks in consumed_message()

**Comment by fpagliughi** [comment]
This will be significantly improved in the upcoming v1.5 release...

First, the internal `thread_queue` is now "closable". When closed, any blocked threads will return immediately. So closing the consumer will close the queue, and any blocked thread will regain control.

The application can then call the client's `bool consumer_closed()` method to check for this state when it get control back. This would normally be used as a signal to exit the consumer application.

Alternately, the consumer queue has been expanded to be an overall client `event` queue which covers the different events that used to be only available through call backs. So instead of just getting messages, it can get alerted to state changes in the client as well. The `event` is a _std::variant_ of different types: *connected, message, connection_lost, disconnected,* and (consumer) *shutdown*. 

These events can be accessed through the `consume_event()` methods in the client.

See the example apps in the master branch for guidance.

**Comment by fpagliughi** [comment]
Should be officially released in a few hours...
